### PR TITLE
Randomized DLS

### LINKED ISSUE #263: Cars All Choose the Same Lane Causing Congestion
> # Temporary solution
> Set DLS to less than 50% in mod options.
----

# New Issue Description
To summarize on what this issue is, as seen in the images below, cars often prefer to use one lane other than a bunch of the other open lanes leading into the same direction. My theory on what's happening? I think cars will prefer to use the shortest distance without taking into account the cars that could block their path (taking more time) resulting in vehicles all choosing the same lane.

I mean, that's pretty much it, the previous description included a lot of debate and conversation about the implications of what's needed to make this happen, but I am 99% sure this is the direction I want to steer this issue in (as it's been all over the place).

---
# Previous Issue Description
This was talked about in the Discord, and I'm putting it here so we can include more people. I'll also be transferring over a lot of the discussion we had in the Discord. Images will be uploaded to the bottom of this issue and will be referenced.

Basically, the idea of this feature is that cars will calculate which lane would be the best lane to go into before entering the next segment. The way it would work is like this:

In Image 2, the blue path represents what was already calculated by the main pathfinder. This is what the car will do if all else fails. The red path represents what the car could do because of this feature. Since the lane is clear, it'd result in a better flow of traffic. Now, in Image 3, let's say the car is currently in segment 0. The pathfinder wants the car to go to segment three. It did this by choosing the middle lane. What this logic would do is it would check the other lanes within segment 1 and determine which lane is the best lane to lead the car to segment 3.

For an example of a failed case, the logic would check the bottommost lane (still referring to Image 3). The bottommost lane goes straight, meaning it wouldn't lead to segment three, therefore being an invalid path.

For an example of a success case, there's actually two situations in this image that could be successful: the bottommost turning lane and the topmost turning lane. Since they both get you to segment three, we move on to the next part of the check. We check both lanes to determine how many vehicles are in both lanes. Since the topmost turning lane has the least amount of vehicles, and it satisfies the first half of the check, the AI would switch the vehicle's path to the topmost turning lane.

As I've stated in the past for other development features (turn on red, emergency vehicle calculation), it would be unbelievably helpful to have a utility class that lists the vehicles within a segment's lane. Code wise, there's multiple situations where we would need this feature to get the vehicles in another lane. For this issue, we could use the calculation to determine how many cars were in each lane. This utility class could be updated by the cars pathfinder's themselves, and when the car is on the main path, it would update the segment (or lane) to say that it's in that lane and remove itself from the old lane. It's simple logic that wouldn't need to be done live, and would save many calculations in many features.

Due to the fact this feature overall would require live calculations, however, it's a very controversial issue and is believed to be a huge performance drag. I don't think so, especially if we implement the utility class, but I'd have to see an implementation of it to fully believe it'll cause major issues; and I mean a well made implementation, not a half-assed put together one. It's not like we haven't done this before with other features that would cause a huge performance drop -- we found utilities, better methods, and good ways to make them more performance-safe. And there's a reason we have _options_, it's to let the user decide which features they'd like to have on or off. If we implement option tooltips that I talked about a little bit with the devs, we could include a warning stating that it could be a performance. Or we could move some other performance-dropping features into a category stating they'll possibly drop performance. I don't think we need to do either of these personally, but some people do, so /shrug.

(Image 1 - The Issue at Hand)
The Issue at Hand

(Image 2 - The Possible Path Calculations; Blue is pre-calculated, red is live-calculated)
The Possible Path Calculations; Blue is pre-calculated, red is live-calculated

(Image 3 - The Segments; Used for Context)
The Segments; Used for Context

### PR DESCRIPTION
**Pull request by VictorPhilipp**
DLS parameters are randomized. Random seed
1. is different for different cims and
2. ~~changes over time.~~

This simulates
1. that different people show different driving styles and
2. ~~that the same person changes their driving style over time~~

Default DLS parameter ranges could/should be optimized.

### CODE REVIEW

**Comment by aubergine10** [comment]
I'll do some testing on this tonight. :D

**Comment by aubergine10** [comment]
Rather than changing over time, could it be changed based on the type of road?

* Highway = frequent
* Large road = common
* Medium road = occasional
* Small road = rare

**Comment by aubergine10** [comment]
Having tested briefly, it looks like it's working, but I wasn't experiencing the same level of problem as FireController so I've requested him to review.

**Comment by aubergine10** [comment]
I'm noticing fps seems lower than usual testing this PR. I normally get 22-25 fps (TMPE v10.20), but with this PR (TMPE v10.18 #334) I'm getting about 18-19 fps.

**Comment by VictorPhilipp** [comment]
> I'm noticing fps seems lower than usual testing this PR
Did you build it with a debug or release configuration? I checked both versions but found not difference in FPS.

> Rather than changing over time, could it be changed based on the type of road?
DLS parameters were designed to model psychologic components of human driving behavior (egoistic vs. altruistic behavior). From your own driving experience: Do you really consider the type of road you are driving on to determine whether a lane change makes sense for you?

I think egoistic drivers will use every opportunity they get to improve their travel time - independent of the road.

> Is there even any need to change over time?
There is no need to change it over time. I wanted to express that even the same person might change driving behavior during a trip. The system would  alter DLS parameters after passing ca.10 junctions on average. I will make it configurable.

**Comment by aubergine10** [comment]
> Did you build it with a debug or release configuration

release

I'll do some more testing as several other mods have been updated this week and one of those could potentially be causing the fps drop.

> Do you really consider the type of road you are driving on to determine whether a lane change makes sense for you?

Yes, absolutely. If I'm on a motorway I'm more likely to overtake than if I'm on a small 2-lane road. Also, time of day also factors into this; if it's late at night, people are generally more inclined to saunter along, even on motorways. And cim age would also be a factor.

Those who use every opportunity to overtake are likely in the realms of reckless driving; putting themselves first to such an extent it's endangering other drivers.

What I'd expect (I realise some or most of this might be completely unrealistic in game performance terms):

* Overtaking to be more prevalent between 7am and 7pm
* Much less overtaking between 11pm and 6am
* Cims travelling to work to be more likely to overtake
* Older cims to be less likely to overtake
* Younger cims to be more likely to overtake
* Overtaking more likely if leader car is going much slower than speed limit (eg. due to slow vehicle asset or 'realistic driving speeds')

**Comment by aubergine10** [comment]
Actualy..

> * Overtaking to be more prevalent between 7am and 7pm
> * Much less overtaking between 11pm and 6am

It's not so much a factor of time, it's more about light intensity, rain/snow intensity, and fog intensity.

Ultimate Eye Candy mod has sliders for the rain and fog intensity, so that would be good reference to find out where those values live. And solar panels state the light intensity, so whatever AI handles solar would be good starting point to locate that value.

Those three factors could be inspected every `X` frames to compute modifiers for a) driving speed and b) overtaking.

**Comment by VictorPhilipp** [comment]
I feel the discussion moves out of scope of what @FireController1847 originally wanted to be adressed.

The original motivation was to prevent DLS-directed lane changes from happening at the same location. In the latest commit I removed the time-varying part. What remains is the per-driver randomization of DLS parameters.

I renamed the feature toggle "Realistic speeds" to "Individual driving styles". It now also controls whether to perform DLS parameter randomization. When disabled, the old constant DLS cost scheme is applied.

**Comment by aubergine10** [comment]
> I feel the discussion moves out of scope

Agreed, I'll create a separate issue to track the other ideas which can be considered at a later date.

> I renamed the feature toggle "Realistic speeds" to "Individual driving styles".

I assume the same pseudo-random seed is used to determine both vehicle speed and DLS application choices?

**Comment by aubergine10** [comment]
@FireController1847 Isn't that separate issue that vehicles only choose the outermost lane when making turns at junctions? I think there was another issue discussing that?

Or it might just be pathfinder going "oh, this is the shortest route, get in this lane"?

**Comment by FireController1847** [comment]
@aubergine10 As seen in the images of #263, that is the *exact* issue I was talking about. Whether or not my issue got changed to something else I'm unaware of, but like I stated I have not noticed any different in terms of lane selection before and after using this PR.

**Comment by aubergine10** [comment]
Just seen a user on Reddit with this problem (somewhat crazy choice of roads)...

l5g18u7js7131

I think it clearly illustrates the issue however: When coming up to a junction, cars are always choosing shortest route. I think the fix to this issue has to also include some sort of "imminent junction" detection and reduce the desirability of shortest route.

I don't know if it's possible, but could we introduce randomisation of which entry lane is tried first at a junction? So if there are X lanes turning in the desired direction, randomise the choice there and if it works don't test other lanes?

**Comment by aubergine10** [comment]
Attempting #316 is going to be insane amount of work and no guarantee that it will even be feasible or solve anything.

I've just added this comment in #263 on what I perceive the issue to be: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/263#issuecomment-497536924

tl;dr: I think individual driving styles are a much needed enhancement to DLS 👍 , but there needs to be separate work taking place on improving pathfinder lane-balancing at junctions 🔨 .

**Comment by VictorPhilipp** [comment]
> I don't know if it's possible, but could we introduce randomisation of which entry lane is tried first at a junction? So if there are X lanes turning in the desired direction, randomise the choice there (regardless of DLS settings) and if it works don't test other lanes?

Advanced Vehicle AI currently randomizes exit lanes for light vehicles. https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/blob/f39ca10311c65db0c67ca05eec423759deafb3f9/TLM/TLM/Custom/PathFinding/CustomPathFind2.cs#L3080-L3103
